plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1584,6 +1584,33 @@ public void addColumn(ConnectorSession session, ConnectorTableHandle tableHandle
         }
     }
 
+    @Override
+    public void addField(ConnectorSession session, ConnectorTableHandle tableHandle, List<String> parentPath, String fieldName, io.trino.spi.type.Type type)
+    {
+        // Iceberg disallows ambiguous field names in a table. e.g. (a row(b int), "a.b" int)
+        String parentName = String.join(".", parentPath);
+
+        Table icebergTable = catalog.loadTable(session, ((IcebergTableHandle) tableHandle).getSchemaTableName());
+        NestedField parent = icebergTable.schema().findField(parentName);
+        verify(parent.type().isStructType(), "Expect struct type");
+        if (parent.type().asStructType().caseInsensitiveField(fieldName) != null) {
+            // TODO https://github.com/trinodb/trino/issues/16273 Adding fields with different cases leads to query failure
+            throw new TrinoException(NOT_SUPPORTED, "This connector does not support adding ambiguous fields");
+        }
+
+        // Start explicitly with highestFieldId + 2 to account for existing columns and the new one being
+        // added - instead of relying on addColumn in iceberg library to assign Ids
+        AtomicInteger nextFieldId = new AtomicInteger(icebergTable.schema().highestFieldId() + 2);
```

Comment:
Why `+2` and not `+1`?
more importantly, what's wrong with "relying on addColumn in iceberg library to assign Ids"?
basing anything on `highestFieldId` seems wrong, since you can "resurrect" fields that previously had highest numbers and got dropped.